File: autoencrypted.go
// Version 2 of autoencrypted files
package iblfile

import (
	"archive/tar"
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
)

var (
	AutoEncryptedFileMagic        = []byte("iblaef")
	AutoEncryptedFileChecksumSize = 32 // sha256
	AutoEncryptedFileIDSize       = 16
)

func AutoEncryptedMetadataSize() int {
	return len(AutoEncryptedFileMagic) + AutoEncryptedFileChecksumSize + AutoEncryptedFileIDSize
}

// Autoencrypted files can be encypted in many ways
//
// This defines an interface for all of them
type AutoEncryptor interface {
	// Returns the identifier of the source, must be unique
	//
	// Max size: 8 ASCII characters (8 bytes)
	ID() string
	// Encrypts a byte slice
	Encrypt([]byte) ([]byte, error)
	// Decrypts a byte slice
	Decrypt([]byte) ([]byte, error) // Decrypts a byte slice
}

var AutoEncryptorRegistry = make(map[string]AutoEncryptor)

func RegisterAutoEncryptor(src AutoEncryptor) {
	id := []byte(src.ID())

	if len(id) != AutoEncryptedFileIDSize {
		panic(fmt.Errorf("invalid id size for %v: %v", src.ID(), len(id)))
	}

	AutoEncryptorRegistry[string(id)] = src
}

// Represents an autoencrypted file block
type AutoEncryptedFileBlock struct {
	// Magic bytes
	Magic []byte
	// Checksum
	Checksum []byte
	// Encryptor
	Encryptor []byte
	// Data
	Data []byte
}

// Validates a block to ensure that it is a valid autoencrypted file block
func (b *AutoEncryptedFileBlock) Validate() error {
	if string(b.Magic) != string(AutoEncryptedFileMagic) {
		return fmt.Errorf("invalid magic: %v", b.Magic)
	}

	// Calculate sha256 checksum of data
	checksum := sha256.Sum256(b.Data)

	if string(checksum[:]) != string(b.Checksum) {
		return fmt.Errorf("invalid checksum: %v", b.Checksum)
	}

	return nil
}

// Decrypts a block into a byte slice
func (b *AutoEncryptedFileBlock) Decrypt(src AutoEncryptor) ([]byte, error) {
	if src.ID() != string(b.Encryptor) {
		return nil, fmt.Errorf("invalid encryptor: %v", b.Encryptor)
	}

	return src.Decrypt(b.Data)
}

// Writes a block to a writer encrypting it with the src
func (b *AutoEncryptedFileBlock) Write(src AutoEncryptor, w io.Writer) error {
	_, err := w.Write(b.Magic)

	if err != nil {
		return err
	}

	if len(b.Magic) != len(AutoEncryptedFileMagic) {
		return fmt.Errorf("magic is not the correct size")
	}

	_, err = w.Write(b.Checksum)

	if err != nil {
		return err
	}

	if len(b.Checksum) != AutoEncryptedFileChecksumSize {
		return fmt.Errorf("checksum is not the correct size")
	}

	_, err = w.Write(b.Encryptor)

	if err != nil {
		return err
	}

	if len(b.Encryptor) != AutoEncryptedFileIDSize {
		return fmt.Errorf("encryptor is not the correct size")
	}

	_, err = w.Write(b.Data)

	if err != nil {
		return err
	}

	return nil
}

func ParseAutoEncryptedFileBlock(block []byte) (*AutoEncryptedFileBlock, error) {
	if len(block) < AutoEncryptedMetadataSize() {
		return nil, fmt.Errorf("block is too small")
	}

	var currentPos int

	// Magic
	magic := block[currentPos : currentPos+len(AutoEncryptedFileMagic)]
	currentPos += len(AutoEncryptedFileMagic)

	// Checksum
	checksum := block[currentPos : currentPos+AutoEncryptedFileChecksumSize]
	currentPos += AutoEncryptedFileChecksumSize

	// Encryptor
	encryptor := block[currentPos : currentPos+AutoEncryptedFileIDSize]
	currentPos += AutoEncryptedFileIDSize

	// Data
	data := block[currentPos:]

	return &AutoEncryptedFileBlock{
		Magic:     magic,
		Checksum:  checksum,
		Encryptor: encryptor,
		Data:      data,
	}, nil
}

func NewAutoEncryptedFileBlock(data []byte, src AutoEncryptor) (*AutoEncryptedFileBlock, error) {
	checksum := sha256.Sum256(data)
	encData, err := src.Encrypt(data)

	if err != nil {
		return nil, err
	}

	return &AutoEncryptedFileBlock{
		Magic:     AutoEncryptedFileMagic,
		Checksum:  checksum[:],
		Encryptor: []byte(src.ID()),
		Data:      encData,
	}, nil
}

// A full file autoencrypted  file. This type stores all data as one single encrypted block rather than per-section blocks
//
// This is the first, and simplest+quickest autoencrypted () file
type AutoEncryptedFile_FullFile struct {
	src      AutoEncryptor
	file     *File
	sections map[string]*bytes.Buffer
}

func NewAutoEncryptedFile_FullFile(src AutoEncryptor) *AutoEncryptedFile_FullFile {
	buf := bytes.NewBuffer([]byte{})
	tarWriter := tar.NewWriter(buf)

	return &AutoEncryptedFile_FullFile{
		src: src,
		file: &File{
			buf:       buf,
			tarWriter: tarWriter,
		},
	}
}

// OpenAutoEncryptedFile_FullFile opens a full file as a single autoencrypted  block
func OpenAutoEncryptedFile_FullFile(r io.Reader, src AutoEncryptor) (*AutoEncryptedFile_FullFile, error) {
	data, err := io.ReadAll(r)

	if err != nil {
		return nil, err
	}

	block, err := ParseAutoEncryptedFileBlock(data)

	if err != nil {
		return nil, err
	}

	if err := block.Validate(); err != nil {
		return nil, fmt.Errorf("block is not valid: %v", err)
	}

	decryptedBlock, err := block.Decrypt(src)

	if err != nil {
		return nil, err
	}

	buf := bytes.NewBuffer(decryptedBlock)
	tarWriter := tar.NewWriter(buf)

	return &AutoEncryptedFile_FullFile{
		src: src,
		file: &File{
			buf:       buf,
			tarWriter: tarWriter,
		},
	}, nil
}

// Returns all sections of the file
func (f *AutoEncryptedFile_FullFile) Sections() (map[string]*bytes.Buffer, error) {
	if f.sections != nil {
		return f.sections, nil
	}

	if f.file.buf.Len() == 0 {
		return map[string]*bytes.Buffer{}, nil
	}

	// Now, we have a decrypted tar file
	files, err := RawDataParse(f.file.buf)

	if err != nil {
		return nil, fmt.Errorf("failed to parse raw data: %w", err)
	}

	f.sections = files
	return files, nil
}

// Get a section from the file
func (f *AutoEncryptedFile_FullFile) Get(name string) (*bytes.Buffer, error) {
	sections, err := f.Sections()

	if err != nil {
		return nil, err
	}

	section, ok := sections[name]

	if !ok {
		return nil, fmt.Errorf("no section found for %s", name)
	}

	return section, nil
}

// Adds a section to a file with json file format
func (f *AutoEncryptedFile_FullFile) WriteJsonSection(i any, name string) error {
	buf := bytes.NewBuffer([]byte{})

	err := json.NewEncoder(buf).Encode(i)

	if err != nil {
		return err
	}

	return f.WriteSection(buf, name)
}

// Adds a section to a file
func (f *AutoEncryptedFile_FullFile) WriteSection(buf *bytes.Buffer, name string) error {
	var err error

	if f.sections == nil {
		f.sections, err = f.Sections()

		if err != nil {
			return err
		}
	}

	err = f.file.WriteSection(buf, name)

	if err != nil {
		return err
	}

	f.sections[name] = buf
	return nil
}

func (f *AutoEncryptedFile_FullFile) WriteOutput(w io.Writer) error {
	dataBuf := bytes.NewBuffer([]byte{})
	err := f.file.WriteOutput(dataBuf)

	if err != nil {
		return err
	}

	encData, err := f.src.Encrypt(dataBuf.Bytes())

	if err != nil {
		return err
	}

	checksum := sha256.Sum256(encData)

	encBlock := AutoEncryptedFileBlock{
		Magic:     AutoEncryptedFileMagic,
		Checksum:  checksum[:],
		Encryptor: []byte(f.src.ID()),
		Data:      encData,
	}

	return encBlock.Write(f.src, w)
}

// Returns the size of the file
func (f *AutoEncryptedFile_FullFile) Size() int {
	return f.file.Size()
}

// The second type of autoencrypted file
//
// This one encrypts individual sections
type AutoEncryptedFile_PerSection struct {
	file *File

	// Unline FullFile, this one stores sections as blocks
	sections map[string]*AutoEncryptedFileBlock
}

func NewAutoEncryptedFile_PerSection() *AutoEncryptedFile_PerSection {
	return &AutoEncryptedFile_PerSection{
		file:     New(),
		sections: make(map[string]*AutoEncryptedFileBlock),
	}
}

// OpenAutoEncryptedFile_PerSection opens a per-section autoencrypted  file
func OpenAutoEncryptedFile_PerSection(r io.Reader, src AutoEncryptor) (*AutoEncryptedFile_PerSection, error) {
	data, err := io.ReadAll(r)

	if err != nil {
		return nil, err
	}

	buf := bytes.NewBuffer(data)

	sections, err := readTarFile(buf)

	if err != nil {
		return nil, err
	}

	encSections := make(map[string]*AutoEncryptedFileBlock)

	for k, v := range sections {
		encSection, err := ParseAutoEncryptedFileBlock(v.Bytes())

		if err != nil {
			return nil, err
		}

		encSections[k] = encSection
	}

	buf = bytes.NewBuffer(data)
	tarWriter := tar.NewWriter(buf)

	return &AutoEncryptedFile_PerSection{
		file: &File{
			buf:       buf,
			tarWriter: tarWriter,
		},
		sections: encSections,
	}, nil
}

// Returns all sections of the file (raw and encrypted)
func (f *AutoEncryptedFile_PerSection) RawSections() (map[string]*AutoEncryptedFileBlock, error) {
	return f.sections, nil
}

// Gets a section given its name and a src
func (f *AutoEncryptedFile_PerSection) Get(name string, src AutoEncryptor) (*bytes.Buffer, error) {
	section, ok := f.sections[name]

	if !ok {
		return nil, fmt.Errorf("no section found for %s", name)
	}

	if err := section.Validate(); err != nil {
		return nil, fmt.Errorf("section is not valid: %v", err)
	}

	decrypted, err := section.Decrypt(src)

	if err != nil {
		return nil, err
	}

	return bytes.NewBuffer(decrypted), nil
}

// Adds a section to a file with json file format
func (f *AutoEncryptedFile_PerSection) WriteJsonSection(i any, name string, src AutoEncryptor) error {
	buf := bytes.NewBuffer([]byte{})

	err := json.NewEncoder(buf).Encode(i)

	if err != nil {
		return err
	}

	return f.WriteSection(buf, name, src)
}

// Adds a section to a file given a buf, a name and a src
func (f *AutoEncryptedFile_PerSection) WriteSection(buf *bytes.Buffer, name string, src AutoEncryptor) error {
	encData, err := src.Encrypt(buf.Bytes())

	if err != nil {
		return err
	}

	checksum := sha256.Sum256(encData)

	encBlock := AutoEncryptedFileBlock{
		Magic:     AutoEncryptedFileMagic,
		Checksum:  checksum[:],
		Encryptor: []byte(src.ID()),
		Data:      encData,
	}

	encBuf := bytes.NewBuffer([]byte{})
	err = encBlock.Write(src, encBuf)

	if err != nil {
		return err
	}

	f.sections[name] = &encBlock
	return f.file.WriteSection(encBuf, name) // We write the encrypted buf to the file
}

func (f *AutoEncryptedFile_PerSection) WriteOutput(w io.Writer) error {
	return f.file.WriteOutput(w)
}

File: rust/src/autoencrypted.rs
use crate::file_format::{read_tar_file, RawFile};
use serde_json;
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::io::{Read, Seek, SeekFrom, Write};

const AUTO_ENCRYPTED_FILE_MAGIC: &[u8] = b"iblaef";
const AUTO_ENCRYPTED_FILE_CHECKSUM_SIZE: usize = 32; // sha256
const AUTO_ENCRYPTED_FILE_ID_SIZE: usize = 16;

fn auto_encrypted_metadata_size() -> usize {
    AUTO_ENCRYPTED_FILE_MAGIC.len()
        + AUTO_ENCRYPTED_FILE_CHECKSUM_SIZE
        + AUTO_ENCRYPTED_FILE_ID_SIZE
}

pub trait AutoEncryptor {
    fn id(&self) -> String;
    fn encrypt(&mut self, data: &[u8]) -> Result<Vec<u8>, crate::Error>;
    fn decrypt(&mut self, data: &[u8]) -> Result<Vec<u8>, crate::Error>;
}

pub struct AutoEncryptorRegistry {
    registry: HashMap<String, Box<dyn AutoEncryptor>>,
}

impl AutoEncryptorRegistry {
    pub fn new() -> Self {
        Self {
            registry: HashMap::new(),
        }
    }

    pub fn register(&mut self, src: Box<dyn AutoEncryptor>) {
        let id = src.id();
        if id.len() != AUTO_ENCRYPTED_FILE_ID_SIZE {
            panic!("invalid id size for {}: {}", id, id.len());
        }
        self.registry.insert(id, src);
    }
}

impl Default for AutoEncryptorRegistry {
    fn default() -> Self {
        Self::new()
    }
}

pub struct AutoEncryptedFileBlock {
    pub magic: Vec<u8>,
    pub checksum: Vec<u8>,
    pub encryptor: Vec<u8>,
    pub data: Vec<u8>,
}

impl AutoEncryptedFileBlock {
    pub fn validate(&self) -> Result<(), crate::Error> {
        if self.magic != AUTO_ENCRYPTED_FILE_MAGIC {
            return Err(format!("invalid magic: {:?}", self.magic).into());
        }

        let mut hasher = Sha256::new();
        hasher.update(&self.data);
        let checksum = hasher.finalize();

        if checksum.to_vec() != self.checksum {
            return Err(format!("invalid checksum: {:?}", self.checksum).into());
        }

        Ok(())
    }

    pub fn decrypt(&self, src: &mut dyn AutoEncryptor) -> Result<Vec<u8>, crate::Error> {
        if src.id() != String::from_utf8_lossy(&self.encryptor) {
            return Err(format!("invalid encryptor: {:?}", self.encryptor).into());
        }

        src.decrypt(&self.data)
    }

    /// Writes a block to a writer with checksum and magic
    pub fn write(&self, w: &mut dyn Write) -> Result<(), crate::Error> {
        w.write_all(&self.magic)?;
        w.write_all(&self.checksum)?;
        w.write_all(&self.encryptor)?;
        w.write_all(&self.data)?;
        Ok(())
    }
}

pub fn parse_auto_encrypted_file_block(
    block: &[u8],
) -> Result<AutoEncryptedFileBlock, crate::Error> {
    if block.len() < auto_encrypted_metadata_size() {
        return Err("block is too small".into());
    }

    let magic = block[0..AUTO_ENCRYPTED_FILE_MAGIC.len()].to_vec();
    let checksum = block[AUTO_ENCRYPTED_FILE_MAGIC.len()
        ..AUTO_ENCRYPTED_FILE_MAGIC.len() + AUTO_ENCRYPTED_FILE_CHECKSUM_SIZE]
        .to_vec();
    let encryptor = block[AUTO_ENCRYPTED_FILE_MAGIC.len() + AUTO_ENCRYPTED_FILE_CHECKSUM_SIZE
        ..AUTO_ENCRYPTED_FILE_MAGIC.len()
            + AUTO_ENCRYPTED_FILE_CHECKSUM_SIZE
            + AUTO_ENCRYPTED_FILE_ID_SIZE]
        .to_vec();
    let data = block[AUTO_ENCRYPTED_FILE_MAGIC.len()
        + AUTO_ENCRYPTED_FILE_CHECKSUM_SIZE
        + AUTO_ENCRYPTED_FILE_ID_SIZE..]
        .to_vec();

    Ok(AutoEncryptedFileBlock {
        magic,
        checksum,
        encryptor,
        data,
    })
}

pub fn new_auto_encrypted_file_block(
    data: &[u8],
    src: &mut dyn AutoEncryptor,
) -> Result<AutoEncryptedFileBlock, crate::Error> {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let checksum = hasher.finalize();

    let enc_data = src.encrypt(data)?;

    Ok(AutoEncryptedFileBlock {
        magic: AUTO_ENCRYPTED_FILE_MAGIC.to_vec(),
        checksum: checksum.to_vec(),
        encryptor: src.id().into_bytes(),
        data: enc_data,
    })
}

pub struct AutoEncryptedFileFullFile {
    src: Box<dyn AutoEncryptor>,
    file: RawFile,
}

impl AutoEncryptedFileFullFile {
    pub fn new(src: Box<dyn AutoEncryptor>) -> Self {
        Self {
            src,
            file: RawFile {
                sections: HashMap::new(),
                size: 0,
            },
        }
    }

    pub fn open(r: &mut dyn Read, src: Box<dyn AutoEncryptor>) -> Result<Self, crate::Error> {
        let mut data = Vec::new();
        r.read_to_end(&mut data)?;

        let block = parse_auto_encrypted_file_block(&data)?;

        block.validate()?;

        let mut src = src;

        let decrypted_block = block.decrypt(&mut *src)?;

        let sections = read_tar_file(&*decrypted_block)?;

        Ok(Self {
            src,
            file: RawFile::from_sections(sections),
        })
    }

    pub fn sections(&self) -> Result<&HashMap<String, Vec<u8>>, crate::Error> {
        Ok(&self.file.sections)
    }

    pub fn get(&mut self, name: &str) -> Result<&Vec<u8>, crate::Error> {
        self.file
            .sections
            .get(name)
            .ok_or_else(|| format!("no section found for {}", name).into())
    }

    pub fn write_json_section<T: serde::Serialize>(
        &mut self,
        item: &T,
        name: &str,
    ) -> Result<(), crate::Error> {
        let buf = serde_json::to_vec(item)?;
        self.write_section(buf, name)
    }

    pub fn write_section(&mut self, buf: Vec<u8>, name: &str) -> Result<(), crate::Error> {
        self.file.write_section(buf, name)
    }

    pub fn write_output(&mut self, w: &mut dyn Write) -> Result<(), crate::Error> {
        let mut data_buf = Vec::new();
        self.file.write_output(&mut data_buf)?;

        let enc_data = self.src.encrypt(&data_buf)?;

        let mut hasher = Sha256::new();
        hasher.update(&enc_data);
        let checksum = hasher.finalize();

        let enc_block = AutoEncryptedFileBlock {
            magic: AUTO_ENCRYPTED_FILE_MAGIC.to_vec(),
            checksum: checksum.to_vec(),
            encryptor: self.src.id().into_bytes(),
            data: enc_data,
        };

        enc_block.write(w)
    }

    pub fn size(&self) -> usize {
        self.file.size()
    }
}

pub struct AutoEncryptedFilePerSection {
    file: RawFile,
    sections: HashMap<String, AutoEncryptedFileBlock>,
    src: Box<dyn AutoEncryptor>,
}

impl AutoEncryptedFilePerSection {
    pub fn new(src: Box<dyn AutoEncryptor>) -> Self {
        Self {
            file: RawFile {
                sections: HashMap::new(),
                size: 0,
            },
            sections: HashMap::new(),
            src,
        }
    }

    pub fn open(r: &mut dyn Read, src: Box<dyn AutoEncryptor>) -> Result<Self, crate::Error> {
        let mut data = Vec::new();
        r.read_to_end(&mut data)?;

        let sections = read_tar_file(&*data)?;

        let mut enc_sections = HashMap::new();

        for (k, v) in sections.iter() {
            let enc_section = parse_auto_encrypted_file_block(v)?;
            enc_sections.insert(k.to_string(), enc_section);
        }

        Ok(Self {
            file: RawFile::from_sections(sections),
            sections: enc_sections,
            src,
        })
    }

    pub fn raw_sections(&self) -> &HashMap<String, AutoEncryptedFileBlock> {
        &self.sections
    }

    pub fn get(&mut self, name: &str) -> Result<Vec<u8>, crate::Error> {
        let section = match self.sections.get(name) {
            Some(section) => section,
            None => return Err(format!("no section found for {}", name).into()),
        };
        section.validate()?;
        let decrypted = section.decrypt(&mut *self.src)?;
        Ok(decrypted)
    }

    pub fn write_json_section<T: serde::Serialize>(
        &mut self,
        item: &T,
        name: &str,
    ) -> Result<(), crate::Error> {
        let buf = serde_json::to_vec(item)?;
        self.write_section(&buf, name)
    }

    pub fn write_section(&mut self, buf: &[u8], name: &str) -> Result<(), crate::Error> {
        let enc_data = self.src.encrypt(buf)?;

        let mut hasher = Sha256::new();
        hasher.update(&enc_data);
        let checksum = hasher.finalize();

        let enc_block = AutoEncryptedFileBlock {
            magic: AUTO_ENCRYPTED_FILE_MAGIC.to_vec(),
            checksum: checksum.to_vec(),
            encryptor: self.src.id().into_bytes(),
            data: enc_data,
        };

        let mut enc_buf = Vec::new();
        enc_block.write(&mut enc_buf)?;

        self.sections.insert(name.to_string(), enc_block);
        self.file.write_section(enc_buf, name)?;
        Ok(())
    }

    pub fn write_output(&mut self, w: &mut dyn Write) -> Result<(), crate::Error> {
        let mut data_buf = Vec::new();
        self.file.write_output(&mut data_buf)?;

        w.write_all(&data_buf)?;

        Ok(())
    }
}

/*
// QuickBlockParser reads the first AutoEncryptedMetadataSize into a buffer and parses it
//
// Note that the block returned by this is *not* valid and is only meant for quick parsing of the encryptor
func QuickBlockParser(r io.ReadSeeker) (*AutoEncryptedFileBlock, error) {
    // Read the first AutoEncryptedMetadataSize into a buffer
    // This is the metadata section
    buf := make([]byte, AutoEncryptedMetadataSize())
    _, err := r.Read(buf)

    if err != nil {
        return nil, fmt.Errorf("error reading metadata: %w", err)
    }

    // This metadata will be 'corrupt', but we just need the encryptor
    meta, err := ParseAutoEncryptedFileBlock(buf)

    if err != nil {
        return nil, fmt.Errorf("error parsing metadata: %w", err)
    }

    // Seek back to start
    _, err = r.Seek(0, 0)

    if err != nil {
        return nil, fmt.Errorf("error seeking back to start of file: %w", err)
    }

    return meta, nil
}
 */

/// QuickBlockParser reads the first AutoEncryptedMetadataSize into a buffer and parses it
///
/// Note that the block returned by this is *not* valid and is only meant for quick parsing of the encryptor
pub fn quick_block_parser<T: Read + Seek>(
    r: &mut T,
) -> Result<AutoEncryptedFileBlock, crate::Error> {
    // Read the first AutoEncryptedMetadataSize into a buffer
    // This is the metadata section
    let mut buf = vec![0u8; auto_encrypted_metadata_size()];
    r.read_exact(&mut buf)?;

    // This metadata will be 'corrupt', but we just need the encryptor
    let meta = parse_auto_encrypted_file_block(&buf)?;

    // Seek back to start
    r.seek(SeekFrom::Start(0))?;

    Ok(meta)
}
